// facturapi/lib/resources/products.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Products {
    constructor(client) {
        this.client = client;
    }
    /**
     * Creates a new product in your organization
     * @param {Object} data - Product data
     * @returns {Promise} Product object
     */
    create(data) {
        return this.client.post('/products', data).then(r => r.data);
    }
    /**
     * Gets a paginated list of products that belong to your organization
     * @param {[Object]} params - Search parameters
     * @returns {Promise} Search results object. The object contains a `data` property with the list of products.
     */
    list(params) {
        return this.client.get('/products', { params: params }).then(r => r.data);
    }
    /**
     * Gets a single product object
     * @param {string} id - Product Id
     * @returns {Promise} Product object
     */
    retrieve(id) {
        if (!id)
            return Promise.reject(new Error('id is required'));
        return this.client.get('/products/' + id).then(r => r.data);
    }
    /**
     * Updates a product
     * @param {string} id - Product Id
     * @param {Object} data - Product data to update
     * @returns {Promise} Updated product
     */
    update(id, data) {
        return this.client.put('/products/' + id, data).then(r => r.data);
    }
    /**
     * Permanently removes a product from your organization.
     * @param {string} id - Product Id
     * @returns {Promise} Deleted product
     */
    del(id) {
        return this.client.delete('/products/' + id).then(r => r.data);
    }
}
exports.default = Products;
